compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -105,6 +105,13 @@ object Parsers {
     def sourcePos(off: Int = in.offset): SourcePosition =
       source atPos Position(off)
 
+    /** in.offset, except if this is at a new line, in which case `lastOffset` is preferred. */
+    def expectedOffset: Int = {
+      val current = sourcePos(in.offset)
+      val last = sourcePos(in.lastOffset)
+      if (current.line != last.line) in.lastOffset else in.offset
```

Comment:
I experienced two weird behaviors in vscode related to zero-length diagnostics https://github.com/Microsoft/vscode/issues/66482 and end-of-line diagnostics https://github.com/Microsoft/vscode/issues/66481 making the current diagnostics render like this

<img width="271" alt="screenshot 2019-01-14 at 16 37 01" src="https://user-images.githubusercontent.com/1408093/51122694-b405b500-181a-11e9-85b4-2efe0c7fd2c1.png">

I was hoping they would render like this instead

![screenshot 2019-01-14 at 14 48 01](https://user-images.githubusercontent.com/1408093/51122725-c2ec6780-181a-11e9-9e4b-7acd5611b9e8.png)

It might be worth reverting the special handling of end-of-line diagnostics, but I'm not sure.